engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java

```diff
@@ -134,13 +138,59 @@ Object doPanicSentinel(
     throw self;
   }
 
-  @Specialization
+  @Specialization(guards = "warnings.hasWarnings(self)")
   Object doWarning(
       VirtualFrame frame,
       State state,
       UnresolvedSymbol symbol,
-      WithWarnings self,
-      Object[] arguments) {
+      Object self,
+      Object[] arguments,
+      @Cached MethodResolverNode methodResolverNode,
+      @CachedLibrary(limit = "10") WarningsLibrary warnings) {
+    Function function =
+        symbol
+            .getScope()
+            .lookupMethodDefinition(
+                EnsoContext.get(this).getBuiltins().warning().getEigentype(), symbol.getName());
```

Comment:
Do I understand correctly that we modify the dispatch mechanism, so that if a value `x` has warnings attached, then for a call `x.f` we will first try to resolve any methods `f` defined on the `Warning` type and only later will fall back to the actual type of `x`?

This is a highly non trivial modification of the dispatch mechanism. We currently use it in an innocent way, because the relevant methods are defined both on `Warning` and on `Any` and we don't expect any overrides.

But I'm suspicious - since there exists a method `Warning.get_all`, if I do:
```
type My_Fancy_Collection
    Value (x:Integer)

    get_all : Vector Integer
    get_all self = [self.x]

main =
    a = My_Fancy_Collection.Value 42
    b = Warning.attach "WARN" <| My_Fancy_Collection.Value 23

	IO.println a.get_all
	IO.println b.get_all
```
will I get
```
[42]
[23]
```
as I'd expect or will I get some unclear behaviour due to `Warning.get_all` shadowing my `get_all`?

Please add a test for this case.

Reply:
I really don't like completely overriding the dispatch mechanism if a value has warnings attached. I don't expect a value with warnings to behave much differently than a regular one.

I don't want to impose any solutions, but maybe instead of redirecting 'all' calls dispatched on a value with warnings through the `Warning` type (which as I noted above may have potentially bad implications), we can just have a list of these special methods? I'd even suggest just creating a map of `methods that need to have warnings from self passed to it`. That would make it a bit less extensible but at the same time more clear which methods are special - and IMO in this case the clarity is the more important goal here.

Reply:
I think we should _definitely_ add a test for the `get_all` case, regardless of the chosen solution. Especially as `get_all` is a pretty common name and the possibility of changing a method declared by the user to one on the warnings module if warnings are attached is something to be avoided.

Then, if we choose to stay with the changed dispatch mechanism, I think it's very important to have some piece of documentation for it. Even if just a comment in the docstring of the `InvokeMethodNode` - but some description that can be found if we ever wonder why dispatch on values-with-warnings behaves the way it does.